### PR TITLE
Sync exporter from Java Agent

### PR DESCRIPTION
**Pull request by heyams**
@trask

### CODE REVIEW

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/resources/META-INF/native-image/com.azure/azure-monitor-opentelemetry-exporter/reflect-config.json [hunk 1]
```diff
@@ -1,121 +0,0 @@
-[
-  {
-    "name": "com.azure.monitor.opentelemetry.exporter.AzureMonitorLogRecordExporter",
-    "allDeclaredMethods": true,
-    "allDeclaredFields": true
-  },
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/resources/META-INF/native-image/com.azure/azure-monitor-opentelemetry-exporter/reflect-config.json:6]: I think need to bring this file back

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/MessageAggregatorTest.java [hunk 2]
```diff
@@ -19,29 +25,28 @@ static void setUp() {
         networkExceptionStats = new OperationLogger(MessageAggregatorTest.class, "intro", 1);
     }
 
-    @Disabled
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/logging/MessageAggregatorTest.java:22]: I think I had to disable this in this repo, don't remember why...

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseTestBase.java [hunk 3]
```diff
@@ -29,12 +29,7 @@
 import reactor.core.publisher.Mono;
 
 import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Date;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Locale;
+import java.util.*;
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseTestBase.java:32]: let's revert this back

File: eng/versioning/external_dependencies.txt [hunk 4]
```diff
@@ -195,15 +195,20 @@ com.microsoft.azure:msal4j-persistence-extension;1.2.0
...
 io.opentelemetry:opentelemetry-exporter-otlp;1.28.0
 io.opentelemetry:opentelemetry-api-logs;1.26.0-alpha
 io.opentelemetry:opentelemetry-sdk-testing;1.28.0
 io.opentelemetry:opentelemetry-sdk-extension-autoconfigure;1.28.0
+io.opentelemetry:opentelemetry-sdk-extension-autoconfigure-spi;1.28.0
 io.opentelemetry.instrumentation:opentelemetry-runtime-telemetry-java8;1.27.0-alpha
 io.opentelemetry.instrumentation:opentelemetry-spring-boot-starter;1.27.0-alpha
 io.projectreactor:reactor-test;3.4.30
+io.github.hakky54:logcaptor;2.9.0
+com.squareup.okio:okio;3.3.0
```
Comment (trask) [inline eng/versioning/external_dependencies.txt:211]: are these two needed?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/pom.xml [hunk 5]
```diff
@@ -61,11 +61,31 @@
...
     <dependency>
       <groupId>com.github.spotbugs</groupId>
       <artifactId>spotbugs-annotations</artifactId>
       <version>4.2.2</version> <!-- {x-version-update;com.github.spotbugs:spotbugs-annotations;external_dependency} -->
     </dependency>
+    <dependency>
+      <groupId>com.google.auto.service</groupId>
+      <artifactId>auto-service</artifactId>
+      <version>1.1.1</version> <!-- {x-version-update;com.google.auto.service:auto-service;external_dependency} -->
+    </dependency>
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/pom.xml:88]: can remove now?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java [hunk 6]
```diff
@@ -329,7 +333,6 @@ private HttpPipeline createHttpPipeline() {
         return new HttpPipelineBuilder()
             .policies(policies.toArray(new HttpPipelinePolicy[0]))
             .httpClient(httpClient)
-                .tracer(new NoopTracer())
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java:332]: I think we want this change (maybe we only made it in this repo?)

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/NoopTracer.java [hunk 7]
```diff
@@ -1,46 +0,0 @@
...
-// Licensed under the MIT License.
-
-package com.azure.monitor.opentelemetry.exporter;
-
-import com.azure.core.util.Context;
-import com.azure.core.util.tracing.Tracer;
-
-import java.util.Objects;
-
-class NoopTracer implements Tracer {
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/NoopTracer.java:11]: ya, we want this

File: eng/versioning/external_dependencies.txt (hunk 4 above)
Comment (heyams) [reply eng/versioning/external_dependencies.txt:211]: used in tests

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseCoordinator.java [hunk 8]
```diff
@@ -58,10 +58,9 @@ public void run() {
             }
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
+        } catch (ThreadDeath td) {
+            throw td;
         } catch (Throwable t) {
-            if (t instanceof Error) {
-                throw (Error) t;
-            }
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseCoordinator.java:64]: check history, this change may have been made in this repo

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseDataFetcher.java [hunk 9]
```diff
@@ -98,20 +98,16 @@ public void prepareQuickPulseDataForSend(String redirectedEndpoint) {
...
                 try (MDC.MDCCloseable ignored = QUICK_PULSE_SEND_ERROR.makeActive()) {
                     logger.error("Quick Pulse failed to prepare data for send", e);
                 }
+            } catch (ThreadDeath td) {
+                throw td;
             } catch (Throwable t2) {
-                if (t2 instanceof Error) {
-                    throw (Error) t2;
-                }
-
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseDataFetcher.java:114]: same for these other ThreadDeath changes

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseCoordinator.java [hunk 10]
```diff
@@ -58,11 +58,12 @@ public void run() {
             }
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
+        } catch (ThreadDeath td) {
+            throw td;
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseCoordinator.java:62]: I think ThreadDeath was removed intentionally in this repo
```suggestion
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseDataFetcher.java [hunk 11]
```diff
@@ -98,21 +98,22 @@ public void prepareQuickPulseDataForSend(String redirectedEndpoint) {
             if (!sendQueue.offer(request)) {
                 logger.verbose("Quick Pulse send queue is full");
             }
+        } catch (ThreadDeath td) {
+            throw td;
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseDataFetcher.java:102]:
```suggestion
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseDataFetcher.java [hunk 12]
```diff
@@ -98,21 +98,22 @@ public void prepareQuickPulseDataForSend(String redirectedEndpoint) {
...
             if (e instanceof Error) {
                 throw (Error) e;
             }
-
             try {
                 try (MDC.MDCCloseable ignored = QUICK_PULSE_SEND_ERROR.makeActive()) {
                     logger.error("Quick Pulse failed to prepare data for send", e);
                 }
+            } catch (ThreadDeath td) {
+                throw td;
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/quickpulse/QuickPulseDataFetcher.java:112]:
```suggestion
```

File: eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml [hunk 13]
```diff
@@ -570,5 +570,4 @@ the main ServiceBusClientBuilder. -->
   <suppress checks="com.azure.tools.checkstyle.checks.ThrowFromClientLoggerCheck" files=".*(DeviceCodeCredential|InteractiveBrowserCredential|AzureCliCredential|AzureDeveloperCliCredential|ClientCertificateCredential|ClientSecretCredential|EnvironmentCredential|OnBehalfOfCredential|ChainedTokenCredential|ClientAssertionCredential|UsernamePasswordCredential).java"/>
 
   <suppress checks="com.azure.tools.checkstyle.checks.EnforceFinalFieldsCheck" files="com.azure.search.documents.indexes.models.SynonymMap"/>
-
```
Comment (trask) [inline eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml:573]: (can undo this change)

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/configuration/StatsbeatConnectionString.java [hunk 14]
```diff
@@ -111,8 +113,8 @@ public static StatsbeatConnectionString create(
             ingestionEndpoint += "/";
         }
         try {
-            endpointUrl = new URL(ingestionEndpoint);
-        } catch (MalformedURLException e) {
+            endpointUrl = new URI(ingestionEndpoint).toURL();
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/configuration/StatsbeatConnectionString.java:116]: what was the warning about using URL constructor? wonder if it applies to URL in general?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineRequest.java [hunk 15]
```diff
@@ -52,16 +53,16 @@ public String getInstrumentationKey() {
     HttpRequest createHttpRequest() {
         HttpRequest request = new HttpRequest(HttpMethod.POST, url);
         request.setBody(Flux.fromIterable(telemetry));
-        request.setHeader("Content-Length", Integer.toString(contentLength));
+        request.setHeader(HttpHeaderName.fromString("Content-Length"), Integer.toString(contentLength));
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineRequest.java:56]: let's make constants for these `HttpHeaderName` values (I _think_ that's the reason for prefering this over passing in String)

**Review by trask — APPROVED** [review]

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/configuration/StatsbeatConnectionString.java (hunk 14 above)
Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/configuration/StatsbeatConnectionString.java:116]: deprecation
